File: examples/sandbox/test-client/src/serde_test.ts
import { AxiosInstance } from "axios"
import {AdjacentlyTaggedEnum, InternallyTaggedEnum, UntaggedEnum} from "./api/things"
import {assert_eq} from "./util";

export default async function run_all(client: AxiosInstance) {
    // untagged enum variant
    await untagged_enum_struct(client)
    await untagged_enum_tuple(client)
    await untagged_enum_tiny_tuple(client)
    await untagged_enum_unit(client)
    await untagged_enum_big_struct(client)

    // adjacently-tagged enum variant
    await adjacent_enum_struct(client)
    await adjacent_enum_tuple(client)
    await adjacent_enum_tiny_tuple(client)
    await adjacent_enum_unit(client)
    await adjacent_enum_big_struct(client)

    // internally-tagged enum variant
    await internal_enum_struct(client)
    await internal_enum_unit(client)
    await internal_enum_big_struct(client)

}

async function untagged_enum_struct(client: AxiosInstance) {
    let result = await client.get<UntaggedEnum>('/untagged_enum_struct')
    console.log("untagged_enum_struct: ", result.data)

    let data: UntaggedEnum = {
        foo: {
            one: 4,
            two: "eight",
        },
        bar: "sixteen",
    }

    let ret = await client.post<UntaggedEnum>('/untagged_enum_struct', data)
    assert_eq(data, ret.data)
}

async function untagged_enum_tuple(client: AxiosInstance) {
    let result = await client.get<UntaggedEnum>('/untagged_enum_tuple')
    console.log("untagged_enum_tuple: ", result.data)

    let data: UntaggedEnum = ["red", "two"]

    let ret = await client.post<UntaggedEnum>('/untagged_enum_tuple', data)
    assert_eq(data, ret.data)
}

async function untagged_enum_tiny_tuple(client: AxiosInstance) {
    let result = await client.get<UntaggedEnum>('/untagged_enum_tiny_tuple')
    console.log("untagged_enum_tiny_tuple: ", result.data)

    let data: UntaggedEnum = "lol"

    let ret = await client.post<UntaggedEnum>('/untagged_enum_tiny_tuple', data)
    assert_eq(data, ret.data)
}

async function untagged_enum_unit(client: AxiosInstance) {
    let result = await client.get<UntaggedEnum>('/untagged_enum_unit')
    console.log("untagged_enum_unit: ", result.data)

    let data: UntaggedEnum = "unit"

    let ret = await client.post<UntaggedEnum>('/untagged_enum_unit', data)
    assert_eq(data, ret.data)
}

async function untagged_enum_big_struct(client: AxiosInstance) {
    let result = await client.get<UntaggedEnum>('/untagged_enum_big_struct')
    console.log("untagged_enum_big_struct: ", result.data)

    let data: UntaggedEnum = {
        THREE: 3,
        six: [0, {one: 1, two: "2"}],
    }

    let ret = await client.post<UntaggedEnum>('/untagged_enum_big_struct', data)
    assert_eq(data, ret.data)
}

async function adjacent_enum_struct(client: AxiosInstance) {
    let result = await client.get<AdjacentlyTaggedEnum>('/adjacent_enum_struct')
    console.log("adjacent_enum_struct: ", result.data)

    let data: AdjacentlyTaggedEnum = {
        t: "Struct",
        c: {
            foo: {
                one: 4,
                two: "eight",
            },
            bar: "sixteen"
        }
    }

    let ret = await client.post<AdjacentlyTaggedEnum>('/adjacent_enum_struct', data)
    assert_eq(data, ret.data)
}

async function adjacent_enum_tuple(client: AxiosInstance) {
    let result = await client.get<AdjacentlyTaggedEnum>('/adjacent_enum_tuple')
    console.log("adjacent_enum_tuple: ", result.data)

    let data: AdjacentlyTaggedEnum = {
        t: "Tuple",
        c: ["red", "two"],
    }

    let ret = await client.post<AdjacentlyTaggedEnum>('/adjacent_enum_tuple', data)
    assert_eq(data, ret.data)
}

async function adjacent_enum_tiny_tuple(client: AxiosInstance) {
    let result = await client.get<AdjacentlyTaggedEnum>('/adjacent_enum_tiny_tuple')
    console.log("adjacent_enum_tiny_tuple: ", result.data)

    let data: AdjacentlyTaggedEnum = {
        t: "TinyTuple",
        c: "lol",
    }

    let ret = await client.post<AdjacentlyTaggedEnum>('/adjacent_enum_tiny_tuple', data)
    assert_eq(data, ret.data)
}

async function adjacent_enum_unit(client: AxiosInstance) {
    let result = await client.get<AdjacentlyTaggedEnum>('/adjacent_enum_unit')
    console.log("adjacent_enum_unit: ", result.data)

    let data: AdjacentlyTaggedEnum = { t: "Unit" }

    let ret = await client.post<AdjacentlyTaggedEnum>('/adjacent_enum_unit', data)
    assert_eq(data, ret.data)
}

async function adjacent_enum_big_struct(client: AxiosInstance) {
    let result = await client.get<AdjacentlyTaggedEnum>('/adjacent_enum_big_struct')
    console.log("adjacent_enum_big_struct: ", result.data)

    let data: AdjacentlyTaggedEnum = {
        t: "BigStruct",
        c: {
            THREE: 3,
            six: [0, {one: 1, two: "2"}],
        }
    }

    let ret = await client.post<AdjacentlyTaggedEnum>('/adjacent_enum_big_struct', data)
    assert_eq(data, ret.data)
}

async function internal_enum_struct(client: AxiosInstance) {
    let result = await client.get<InternallyTaggedEnum>('/internal_enum_struct')
    console.log("internal_enum_struct: ", result.data)

    let data: InternallyTaggedEnum = {
        type: "Struct",
        foo: {
            one: 4,
            two: "eight",
        },
        bar: "sixteen"
    }

    let ret = await client.post<InternallyTaggedEnum>('/internal_enum_struct', data)
    assert_eq(data, ret.data)
}

async function internal_enum_unit(client: AxiosInstance) {
    let result = await client.get<InternallyTaggedEnum>('/internal_enum_unit')
    console.log("internal_enum_unit: ", result.data)

    let data: InternallyTaggedEnum = { type: "Unit" }

    let ret = await client.post<InternallyTaggedEnum>('/internal_enum_unit', data)
    assert_eq(data, ret.data)
}

async function internal_enum_big_struct(client: AxiosInstance) {
    let result = await client.get<InternallyTaggedEnum>('/internal_enum_big_struct')
    console.log("internal_enum_big_struct: ", result.data)

    let data: InternallyTaggedEnum = {
        type: "BigStruct",
        one: 1,
        two: "two",
        THREE: 3,
        six: [0, {one: 1, two: "2"}],
    }

    let ret = await client.post<InternallyTaggedEnum>('/internal_enum_big_struct', data)
    assert_eq(data, ret.data)
}

File: examples/sandbox/test-server/src/serde_test.rs
use crate::post_echo;
use sandbox::Foo;
use sandbox::inner::bar::DeepTupleStruct;
use sandbox::inner::TupleStruct;
use sandbox::serde_test::*;
use axum::{Json, Router};
use axum::routing::get;

pub fn setup(router: Router) -> Router {
    router
        // untagged enum variant
        .route("/untagged_enum_struct", get(get_untagged_enum_struct).post(post_echo::<UntaggedEnum>))
        .route("/untagged_enum_tuple", get(get_untagged_enum_tuple).post(post_echo::<UntaggedEnum>))
        .route("/untagged_enum_tiny_tuple", get(get_untagged_enum_tiny_tuple).post(post_echo::<UntaggedEnum>))
        .route("/untagged_enum_unit", get(get_untagged_enum_unit).post(post_echo::<UntaggedEnum>))
        .route("/untagged_enum_big_struct", get(get_untagged_enum_big_struct).post(post_echo::<UntaggedEnum>))

        // adjacently-tagged enum variant
        .route("/adjacent_enum_struct", get(get_adjacent_enum_struct).post(post_echo::<AdjacentlyTaggedEnum>))
        .route("/adjacent_enum_tuple", get(get_adjacent_enum_tuple).post(post_echo::<AdjacentlyTaggedEnum>))
        .route("/adjacent_enum_tiny_tuple", get(get_adjacent_enum_tiny_tuple).post(post_echo::<AdjacentlyTaggedEnum>))
        .route("/adjacent_enum_unit", get(get_adjacent_enum_unit).post(post_echo::<AdjacentlyTaggedEnum>))
        .route("/adjacent_enum_big_struct", get(get_adjacent_enum_big_struct).post(post_echo::<AdjacentlyTaggedEnum>))

        // internally-tagged enum variant
        .route("/internal_enum_struct", get(get_internal_enum_struct).post(post_echo::<InternallyTaggedEnum>))
        .route("/internal_enum_unit", get(get_internal_enum_unit).post(post_echo::<InternallyTaggedEnum>))
        .route("/internal_enum_big_struct", get(get_internal_enum_big_struct).post(post_echo::<InternallyTaggedEnum>))
}

async fn get_untagged_enum_struct() -> Json<UntaggedEnum> {
    let data = UntaggedEnum::Struct {
        foo: Foo { one: 8, two: "16".to_string() },
        bar: "bar".to_string(),
    };

    Json(data)
}

async fn get_untagged_enum_tuple() -> Json<UntaggedEnum> {
    let data = UntaggedEnum::Tuple(Stuff::Two, Stuff::Red);

    Json(data)
}

async fn get_untagged_enum_tiny_tuple() -> Json<UntaggedEnum> {
    let data = UntaggedEnum::TinyTuple("kek".into());

    Json(data)
}

async fn get_untagged_enum_unit() -> Json<UntaggedEnum> {
    let data = UntaggedEnum::Unit;

    Json(data)
}

async fn get_untagged_enum_big_struct() -> Json<UntaggedEnum> {
    let data = UntaggedEnum::BigStruct {
        three: DeepTupleStruct(4),
        four: Some(NamedStruct {
            foo: Stuff::Red,
            opt: None,
            more: Foo {
                one: 42,
                two: "24".to_string(),
            },
        }),
        five: TupleStruct(8, Foo {
            one: 16,
            two: "32".to_string(),
        }),
    };

    Json(data)
}

async fn get_adjacent_enum_struct() -> Json<AdjacentlyTaggedEnum> {
    let data = AdjacentlyTaggedEnum::Struct {
        foo: Foo { one: 8, two: "16".to_string() },
        bar: "bar".to_string(),
    };

    Json(data)
}

async fn get_adjacent_enum_tuple() -> Json<AdjacentlyTaggedEnum> {
    let data = AdjacentlyTaggedEnum::Tuple(Stuff::Red, Stuff::Two);

    Json(data)
}

async fn get_adjacent_enum_tiny_tuple() -> Json<AdjacentlyTaggedEnum> {
    let data = AdjacentlyTaggedEnum::TinyTuple("kek".into());

    Json(data)
}

async fn get_adjacent_enum_unit() -> Json<AdjacentlyTaggedEnum> {
    let data = AdjacentlyTaggedEnum::Unit;

    Json(data)
}

async fn get_adjacent_enum_big_struct() -> Json<AdjacentlyTaggedEnum> {
    let data = AdjacentlyTaggedEnum::BigStruct {
        three: DeepTupleStruct(4),
        four: Some(NamedStruct {
            foo: Stuff::Red,
            opt: None,
            more: Foo {
                one: 42,
                two: "24".to_string(),
            },
        }),
        five: TupleStruct(8, Foo {
            one: 16,
            two: "32".to_string(),
        }),
    };

    Json(data)
}

async fn get_internal_enum_struct() -> Json<InternallyTaggedEnum> {
    let data = InternallyTaggedEnum::Struct {
        foo: Foo { one: 8, two: "16".to_string() },
        bar: "bar".to_string(),
    };

    Json(data)
}

async fn get_internal_enum_unit() -> Json<InternallyTaggedEnum> {
    let data = InternallyTaggedEnum::Unit;

    Json(data)
}

async fn get_internal_enum_big_struct() -> Json<InternallyTaggedEnum> {
    let data = InternallyTaggedEnum::BigStruct {
        more: Foo {
            one: 1,
            two: "2".to_string(),
        },
        three: DeepTupleStruct(4),
        four: Some(NamedStruct {
            foo: Stuff::Red,
            opt: None,
            more: Foo {
                one: 42,
                two: "24".to_string(),
            },
        }),
        five: TupleStruct(8, Foo {
            one: 16,
            two: "32".to_string(),
        }),
    };

    Json(data)
}